cartridge/hotreload.lua

```diff
@@ -0,0 +1,194 @@
+        if not t1[k] then
+            table.insert(ret, k)
+        end
+    end
+
+    table.sort(ret)
+    return ret
+end
+
+--- Avoid removing globals with given names.
+--
+-- @function whitelist_globals
+-- @tparam {string,...} names
+local function whitelist_globals(tbl)
+    if not vars.globals then
+        return
+    end
+
+    for _, v in ipairs(tbl) do
+        log.debug('Avoid removing global %q on hot-reload', v)
+        vars.globals[v] = true
+    end
+end
+
+--- Avoid cancelling fibers with given names.
+--
+-- Names 'lua', 'main' and '* (net.box)' are ignored and can't be
+-- whitelisted.
+--
+-- @function whitelist_fibers
+-- @tparam {string,...} names
+local function whitelist_fibers(tbl)
+    if not vars.fibers then
+        return
+    end
+
+    for _, v in ipairs(tbl) do
+        if v == 'lua'
+        or v == 'main'
+        or v:endswith('(net.box)')
+        then
+            log.debug('Refusing to whitelist fiber %q', v)
+        else
+            log.debug('Avoid cancelling fiber %q on hot-reload', v)
+            vars.fibers[v] = true
+        end
+    end
+end
+
+--- Check that state was saved.
+--
+-- @function state_saved
+-- @treturn boolean true / false
+local function state_saved()
+    return vars.packages
+        and vars.globals
+        and vars.fibers
+        and vars.routes_count
+end
+
+--- Save initial state.
+--
+-- This includes `package.loaded`, `_G`, fibers, http routes.
+--
+-- @function save_state
+local function save_state()
+    vars.packages = table.copy(package.loaded)
+    for k, _ in pairs(vars.packages) do
+        vars.packages[k] = true
+    end
+
+    vars.globals = setmetatable(table.copy(_G), nil)
+    for k, _ in pairs(vars.globals) do
+        vars.globals[k] = true
+    end
+
+    vars.fibers = {}
+    whitelist_fibers(diff(vars.fibers, snap_fibers()))
+
+    local httpd = service_registry.get('httpd')
+    if httpd ~= nil then
+        vars.routes_count = #httpd.routes
+    else
+        vars.routes_count = 0
+    end
+
+    assert(state_saved())
+end
+
+--- Restore initial state.
+--
+-- Unload packages, cancel fibers, remove globals and http routes.
+--
+-- @function load_state
+local function load_state()
+    assert(state_saved(), "Hot-reload state wasn't saved")
+
+    for k, _ in pairs(package.loaded) do
+        if not vars.packages[k] then
+            -- log.info('Unloading package %q', k)
+            package.loaded[k] = nil
+        end
+    end
+
+    for k, _ in pairs(_G) do
+        if not vars.globals[k] then
+            log.info('Unsetting global %q', k)
+            _G[k] = nil
+        end
+    end
+
+    local fiber_info = fiber.info()
+    for _, f in pairs(fiber_info) do
+        if f.fid == fiber.id() then
+            -- don't kill self
+            goto continue
+        end
+
+        if f.fid == 101
+        or f.name:startswith('console/')
+        or f.name:startswith('applier/')
+        or f.name:startswith('applierw/')
```

Comment:
And "raft_worker" https://github.com/tarantool/tarantool/blob/71377c28e1c20108a8691481660bea6263f9a2e8/src/box/raft.c#L120 ?

Reply:
"checkpoint_daemon"? https://github.com/tarantool/tarantool/blob/b8d3c762b284c36c7ba85f1dc4d4607c594e4ffb/src/box/gc.c#L121

Reply:
I think you should revise list of system fibers once again

Reply:
I'm sure `checkpoint_daemon` isn't necessary here. Not sure about the `raft_worker` though. This list is only for system fibers with custom (formatted) names. All fibers spawned during `box.cfg()` are already whitelisted. But the `applier/*` fibers are started while ConnectingFullmesh.

Reply:
I'm not sure it's so
```
tarantool> 2020-12-16 11:11:22.433 [76140] main/116/checkpoint_daemon I> scheduled next checkpoint for Wed Dec 16 11:12:22 2020
Segmentation fault
  code: SEGV_MAPERR
  addr: 0x0
  context: 0x115bffee0
  siginfo: 0x115bffe78
Current time: 1608106282
Please file a bug at http://github.com/tarantool/tarantool/issues
Attempting backtrace... Note: since the server has already crashed, 
this may fail as well
#0  0x10eb37399 in print_backtrace+9
#1  0x10e9df871 in _ZL12sig_fatal_cbiP9__siginfoPv+151
#2  0x7fff75fd2b5d in _sigtramp+1d
[1]    76140 abort (core dumped)  ./init.lua
```

I tried to cancel some fibers and finally got coredump. Seems "gc" is the name of fiber I cancelled.

Reply:
```
tarantool> f.find(115)
---
- status: suspended
  name: gc
  id: 115
...

tarantool> f.find(115):cancel()
---
...

tarantool> f.find(115)
---
- null
...

tarantool> box.snapshot()
2020-12-16 11:25:22.698 [77156] main/116/checkpoint_daemon I> scheduled next checkpoint for Wed Dec 16 13:08:32 2020
2020-12-16 11:25:22.699 [77156] snapshot/101/main I> saving snapshot `/Users/o.babin/Projects/tdg/dev/output/srv/00000000000000030768.snap.inprogress'
2020-12-16 11:25:22.708 [77156] snapshot/101/main I> done
Segmentation fault
  code: SEGV_MAPERR
  addr: 0x0
  context: 0x1127ffee0
  siginfo: 0x1127ffe78
Current time: 1608107122
Please file a bug at http://github.com/tarantool/tarantool/issues
Attempting backtrace... Note: since the server has already crashed, 
this may fail as well
#0  0x10b05a399 in print_backtrace+9
#1  0x10af02871 in _ZL12sig_fatal_cbiP9__siginfoPv+151
#2  0x7fff75fd2b5d in _sigtramp+1d
[1]    77156 abort (core dumped)  ./init.lua
```

Please revise this whilelist and blacklist.

Reply:
```
tarantool> require('cartridge.roles').reload()
2020-12-16 11:39:00.152 [78023] main/103/init.lua roles.lua:440 W> Reloading roles ...
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> Instance state changed: RolesConfigured -> ReloadingRoles
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> Starting reconfiguration of replica 240f306b-bd29-41bc-a8d7-6f6697ece00f
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> Resigning from the replicaset master role...
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> Box has been configured
2020-12-16 11:39:00.152 [78023] main/204/lua I> Old replicaset and replica objects are outdated.
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> GC stopped
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> Recovery stopped
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> Resigned from the replicaset master role
2020-12-16 11:39:00.152 [78023] main/103/init.lua I> Rebalancer location has changed to nil
2020-12-16 11:39:00.168 [78023] main/103/init.lua I> Starting router configuration
2020-12-16 11:39:00.168 [78023] main/103/init.lua I> Calling box.cfg()...
2020-12-16 11:39:00.168 [78023] main/103/init.lua I> Box has been configured
2020-12-16 11:39:00.168 [78023] main/205/lua I> Old replicaset and replica objects are outdated.
2020-12-16 11:39:00.170 [78023] main/103/init.lua I> Unsetting global "vshard_blob_storage"
2020-12-16 11:39:00.170 [78023] main/103/init.lua I> Unsetting global "call_service"
2020-12-16 11:39:00.170 [78023] main/103/init.lua I> Unsetting global "execute_graphql"
2020-12-16 11:39:00.170 [78023] main/103/init.lua I> Unsetting global "smtp"
2020-12-16 11:39:00.170 [78023] main/103/init.lua I> Unsetting global "vshard_sandbox"
2020-12-16 11:39:00.170 [78023] main/103/init.lua I> Unsetting global "vshard_tasks"
2020-12-16 11:39:00.170 [78023] main/103/init.lua I> Unsetting global "vshard_proxy"
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving system fiber "lua" (101)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "on_shutdown" (102)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "common_log" (104)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "common_log_resend" (105)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "audit_log" (106)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "audit_log_resend" (107)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "membership.main" (108)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "membership.entropy" (109)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "membership.handle" (110)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "http/0.0.0.0:8080" (111)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "gc" (115)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "checkpoint_daemon" (116)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "memtx.gc" (117)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "vinyl.scheduler" (118)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "vinyl.vylog_flusher" (119)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "main" (120)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "guard of feedback_daemon" (125)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving whitelisted fiber "feedback_daemon" (126)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Preserving system fiber "applier/admin@localhost:3301" (127)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "vshard.gc" (128)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "vshard.recovery" (129)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "vshard.rebalancer" (130)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "localhost:3301 (net.box)" (131)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "localhost:3301 (net.box)" (132)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "vshard.failover._static_router" (133)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "vshard.discovery._static_router" (134)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: storage.output_replication_queue.object_sender" (135)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: storage.jobs.jobs.job_dispatcher_loop" (136)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: storage.commands.execute" (137)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (138)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (139)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (140)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (141)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (142)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (143)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (144)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: connector.filewatcher.handle_obj_async" (145)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "task: input_processor.repair_queue.repair_queue.resend_objects" (146)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_1" (147)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_2" (148)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_3" (149)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_4" (150)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_5" (151)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_6" (152)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_7" (153)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_8" (154)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_9" (155)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_10" (156)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_11" (157)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_12" (158)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_13" (159)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_14" (160)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_15" (161)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_16" (162)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_17" (163)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_18" (164)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_19" (165)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_20" (166)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_21" (167)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_22" (168)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_23" (169)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_24" (170)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_25" (171)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_26" (172)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_27" (173)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_28" (174)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_29" (175)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_30" (176)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_31" (177)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_32" (178)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_33" (179)
2020-12-16 11:39:00.172 [78023] main/103/init.lua I> Killing fiber "runner_worker_34" (180)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_35" (181)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_36" (182)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_37" (183)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_38" (184)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_39" (185)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_40" (186)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_41" (187)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_42" (188)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_43" (189)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_44" (190)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_45" (191)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_46" (192)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_47" (193)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_48" (194)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_49" (195)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "runner_worker_50" (196)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "ping_subscribers" (197)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "task: account_manager.expiration.server.check_expiration" (198)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "cron" (199)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "cluster_time_checker" (201)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "watchdog_petting" (202)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Killing fiber "watchdog_timer" (203)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/data/:type_name" (PUT)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/data/:type_name" (DELETE)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/data/:type_name" (POST)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/data/:type_name" (GET)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/service/:service_name" (POST)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/soap" (GET)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/soap" (POST)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/http" (POST)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/welcome" (ANY)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/admin/tdg/docs/.*" (ANY)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/metrics" (ANY)
2020-12-16 11:39:00.173 [78023] main/103/init.lua I> Removing HTTP route "/graphql" (POST)
2020-12-16 11:39:00.190 [78023] main/103/init.lua confapplier.lua:131 E> Instance entering failed state: ReloadingRoles -> ReloadError
RegisterRoleError: /Users/o.babin/Projects/tdg/common/document/key_def.lua:3: attempt to redefine 'field_type' at line 42
stack traceback:
	/Users/o.babin/Projects/tdg/common/document/key_def.lua:3: in main chunk
	[C]: in function 'require'
	/Users/o.babin/Projects/tdg/common/document/metrics.lua:12: in main chunk
	[C]: in function 'require'
	/Users/o.babin/Projects/tdg/common/model_accessor.lua:14: in main chunk
	[C]: in function 'require'
	/Users/o.babin/Projects/tdg/storage/expiration/init.lua:6: in main chunk
	[C]: in function 'require'
	/Users/o.babin/Projects/tdg/storage/storage.lua:8: in main chunk
	...
	[C]: in function 'xpcall'
	...s/o.babin/Projects/tdg/.rocks/share/tarantool/errors.lua:145: in function 'pcall'
	.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:49: in function 'register_role'
	.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:151: in function 'cfg'
	.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:460: in function <.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:421>
	[C]: in function 'pcall'
	builtin/box/console.lua:402: in function 'eval'
	builtin/box/console.lua:708: in function 'repl'
	builtin/box/console.lua:757: in function 'start'
	./init.lua:175: in main chunk
---
- null
- line: 3
  class_name: RegisterRoleError
  err: '/Users/o.babin/Projects/tdg/common/document/key_def.lua:3: attempt to redefine
    ''field_type'' at line 42'
  file: /Users/o.babin/Projects/tdg/common/document/key_def.lua
  stack: "stack traceback:\n\t/Users/o.babin/Projects/tdg/common/document/key_def.lua:3:
    in main chunk\n\t[C]: in function 'require'\n\t/Users/o.babin/Projects/tdg/common/document/metrics.lua:12:
    in main chunk\n\t[C]: in function 'require'\n\t/Users/o.babin/Projects/tdg/common/model_accessor.lua:14:
    in main chunk\n\t[C]: in function 'require'\n\t/Users/o.babin/Projects/tdg/storage/expiration/init.lua:6:
    in main chunk\n\t[C]: in function 'require'\n\t/Users/o.babin/Projects/tdg/storage/storage.lua:8:
    in main chunk\n\t...\n\t[C]: in function 'xpcall'\n\t...s/o.babin/Projects/tdg/.rocks/share/tarantool/errors.lua:145:
    in function 'pcall'\n\t.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:49:
    in function 'register_role'\n\t.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:151:
    in function 'cfg'\n\t.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:460:
    in function <.../Projects/tdg/.rocks/share/tarantool/cartridge/roles.lua:421>\n\t[C]:
    in function 'pcall'\n\tbuiltin/box/console.lua:402: in function 'eval'\n\tbuiltin/box/console.lua:708:
    in function 'repl'\n\tbuiltin/box/console.lua:757: in function 'start'\n\t./init.lua:175:
    in main chunk"
  str: 'RegisterRoleError: /Users/o.babin/Projects/tdg/common/document/key_def.lua:3:
    attempt to redefine ''field_type'' at line 42'
...

tarantool> 2020-12-16 11:39:00.190 [78023] main/131/localhost:3301 (net.box) I> disconnected from localhost:3301
2020-12-16 11:39:00.191 [78023] main/132/localhost:3301 (net.box) I> disconnected from localhost:3301
2020-12-16 11:39:00.191 [78023] main/138/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled
2020-12-16 11:39:00.191 [78023] main/139/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled
2020-12-16 11:39:00.191 [78023] main/140/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled
2020-12-16 11:39:00.191 [78023] main/141/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled
2020-12-16 11:39:00.191 [78023] main/142/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled
2020-12-16 11:39:00.191 [78023] main/143/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled
2020-12-16 11:39:00.191 [78023] main/144/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled
2020-12-16 11:39:00.191 [78023] main/145/task: connector.filewatcher.handle_obj_async log.lua:202 E> fiber is cancelled

---
...

tarantool> 
---
...

tarantool> box.snapshot()
2020-12-16 11:39:05.144 [78023] main/116/checkpoint_daemon I> scheduled next checkpoint for Wed Dec 16 12:46:00 2020
2020-12-16 11:39:05.145 [78023] snapshot/101/main I> saving snapshot `/Users/o.babin/Projects/tdg/dev/output/srv/00000000000000030776.snap.inprogress'
2020-12-16 11:39:05.151 [78023] snapshot/101/main I> done
2020-12-16 11:39:05.152 [78023] main I> removed /Users/o.babin/Projects/tdg/dev/output/srv/00000000000000030760.snap
---
- ok
...

tarantool> 2020-12-16 11:39:05.153 [78023] main I> removed /Users/o.babin/Projects/tdg/dev/output/srv/00000000000000000000.vylog

---
...

tarantool> Segmentation fault
  code: SEGV_MAPERR
  addr: 0x0
  context: 0x12047fee0
  siginfo: 0x12047fe78
Current time: 1608107946
Please file a bug at http://github.com/tarantool/tarantool/issues
Attempting backtrace... Note: since the server has already crashed, 
this may fail as well
#0  0x10ff2a399 in print_backtrace+9
#1  0x10fdd2871 in _ZL12sig_fatal_cbiP9__siginfoPv+151
#2  0x7fff75fd2b5d in _sigtramp+1d
[1]    78023 abort (core dumped)  ./init.lua
```